playback/src/audio_backend/alsa.rs

```diff
@@ -106,302 +119,100 @@ fn list_compatible_devices() -> SinkResult<()> {
-    println!("\t------------------------------------------------------\n");
+    println!("\t--------------------------------------------------------------------\n");
 
     for a in i {
         if let Some(Direction::Playback) = a.direction {
             if let Some(name) = a.name {
-                if let Ok(pcm) = PCM::new(&name, Direction::Playback, false) {
-                    if let Ok(hwp) = HwParams::any(&pcm) {
-                        // Only show devices that support
-                        // 2 ch 44.1 Interleaved.
-
-                        if hwp.set_access(Access::RWInterleaved).is_ok()
-                            && hwp.set_rate(SAMPLE_RATE, ValueOr::Nearest).is_ok()
-                            && hwp.set_channels(NUM_CHANNELS as u32).is_ok()
-                        {
-                            let mut supported_formats = vec![];
-
-                            for f in &[
-                                AudioFormat::S16,
-                                AudioFormat::S24,
-                                AudioFormat::S24_3,
-                                AudioFormat::S32,
-                                AudioFormat::F32,
-                                AudioFormat::F64,
-                            ] {
-                                if hwp.test_format(Format::from(*f)).is_ok() {
-                                    supported_formats.push(format!("{f:?}"));
+                // surround* outputs throw:
+                // ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
+                if name.contains(':') && !name.starts_with("surround") {
+                    if let Ok(pcm) = PCM::new(&name, Direction::Playback, false) {
+                        if let Ok(hwp) = HwParams::any(&pcm) {
+                            if hwp.set_access(Access::RWInterleaved).is_ok()
+                                && hwp.set_channels(NUM_CHANNELS as u32).is_ok()
+                            {
+                                let mut supported_formats_and_samplerates = String::new();
+
+                                for format in FORMATS.iter() {
+                                    let hwp = hwp.clone();
+
+                                    if hwp.set_format((*format).into()).is_ok() {
+                                        let sample_rates: Vec<String> = SUPPORTED_SAMPLE_RATES
+                                            .iter()
+                                            .filter_map(|sample_rate| {
+                                                let hwp = hwp.clone();
+                                                if hwp
+                                                    .set_rate(*sample_rate, ValueOr::Nearest)
+                                                    .is_ok()
+                                                {
+                                                    match *sample_rate {
+                                                        44100 => Some("44.1kHz".to_string()),
```

Comment:
If you'd define the sample rates not as an array of integers, but as a struct, then you could use nice `ToString` and `FromStr` and `From<SampleRate> to u32` idioms.

Reply:
An enum you mean?

Reply:
I just impl'd IntoIter for SampleRate and AudioFormat. That seemed to be the easiest way to do what you want I think?

Reply:
I’m on mobile so maybe I missed it, but where’s that impl precisely?

Reply:
In playback config.

I added:

```rust
impl IntoIterator for AudioFormat {
    type Item = AudioFormat;
    type IntoIter = std::vec::IntoIter<Self::Item>;

    fn into_iter(self) -> Self::IntoIter {
        use AudioFormat::*;

        vec![F64, F32, S32, S24, S24_3, S16].into_iter()
    }
}
```

And:

```rust
impl IntoIterator for SampleRate {
    type Item = SampleRate;
    type IntoIter = std::vec::IntoIter<Self::Item>;

    fn into_iter(self) -> Self::IntoIter {
        use SampleRate::*;

        vec![Hz44100, Hz48000, Hz88200, Hz96000].into_iter()
    }
}
```